core/src/main/java/org/elasticsearch/search/aggregations/AggregationBuilder.java

```diff
@@ -101,7 +101,7 @@ public final AggregationBuilder rewrite(QueryRewriteContext context) throws IOEx
         if (rewritten == this) {
             return rewritten;
         }
-        if (getMetaData() != null && rewritten.getMetaData() == null) {
+        if (getMetaData() != null && (rewritten.getMetaData() == null || rewritten.getMetaData().isEmpty())) {
```

Comment:
Shouldn't the agg be responsible for copying the metadata when it's rewritten ? This is a bit confusing now since we don't know if the rewritten metadata is empty on purpose or not. We could just copy the metadata explicitly in `Filter(s)Agg...#doRewrite` to avoid ambiguity ?

Reply:
The way I have been thinking about it is that this superclass manages the metadata in every other respect (toXContent, setters, getters) so it seem natural to me that it would handle the metadata while rewriting the agg? This way the way that metadata is handled should be consistent across all aggregations?

Reply:
Actually it is a bit confusing because `AbractAggregationBuilder` handles some of the metadata stuff too, but since most if not all aggregations extend AbstractAggregationBuilder I think it is still nice to have consistent logic for metadata across all the aggs. Maybe separately we should look at whether we still need AggregationBuilder and AbstractAggregationBuilder or if we could merge them to make things a bit simpler?

Reply:
> Maybe separately we should look at whether we still need AggregationBuilder and AbstractAggregationBuilder or if we could merge them to make things a bit simpler?

agreed, it will also simplify the fix for https://github.com/elastic/elasticsearch/issues/27782 which requires a new clone function.